pkg/common_controller/snapshot_controller.go

```diff
@@ -502,362 +645,270 @@ func (ctrl *csiSnapshotController) isVolumeBeingCreatedFromSnapshot(snapshot *cr
-		metav1.SetMetaDataAnnotation(&snapshotContent.ObjectMeta, AnnDeletionSecretRefNamespace, snapshotterSecretRef.Namespace)
-	}
+	// Copy the snapshot object before updating it
+	snapshotCopy := snapshotObj.DeepCopy()
 
-	klog.V(3).Infof("volume snapshot content %v", snapshotContent)
-	// Try to create the VolumeSnapshotContent object several times
-	for i := 0; i < ctrl.createSnapshotContentRetryCount; i++ {
-		klog.V(5).Infof("createSnapshot [%s]: trying to save volume snapshot content %s", snapshotKey(snapshot), snapshotContent.Name)
-		if _, err = ctrl.clientset.SnapshotV1beta1().VolumeSnapshotContents().Create(snapshotContent); err == nil || apierrs.IsAlreadyExists(err) {
-			// creation succeeded.
-			if err != nil {
-				klog.V(3).Infof("volume snapshot content %q for snapshot %q already exists, reusing", snapshotContent.Name, snapshotKey(snapshot))
-				err = nil
-			} else {
-				klog.V(3).Infof("volume snapshot content %q for snapshot %q saved, %v", snapshotContent.Name, snapshotKey(snapshot), snapshotContent)
-			}
-			newContent, err := ctrl.clientset.SnapshotV1beta1().VolumeSnapshotContents().Get(snapshotContent.Name, metav1.GetOptions{})
-			if err == nil {
-				err = ctrl.updateSnapshotContentStatus(newContent, snapshotID, readyToUse, timestamp, size)
-			}
-			break
+	if snapshot.Status != nil && snapshot.Status.BoundVolumeSnapshotContentName != nil && *snapshotObj.Status.BoundVolumeSnapshotContentName == snapshotContent.Name {
+		klog.Infof("bindVolumeSnapshotContentToVolumeSnapshot: VolumeSnapshot %s already bind to volumeSnapshotContent [%s]", snapshot.Name, snapshotContent.Name)
+	} else {
+		klog.Infof("bindVolumeSnapshotContentToVolumeSnapshot: before bind VolumeSnapshot %s to volumeSnapshotContent [%s]", snapshot.Name, snapshotContent.Name)
+		if snapshotCopy.Status == nil {
+			snapshotCopy.Status = &crdv1.VolumeSnapshotStatus{}
 		}
-		// Save failed, try again after a while.
-		klog.V(3).Infof("failed to save volume snapshot content %q for snapshot %q: %v", snapshotContent.Name, snapshotKey(snapshot), err)
-		time.Sleep(ctrl.createSnapshotContentInterval)
-	}
+		snapshotCopy.Status.BoundVolumeSnapshotContentName = &snapshotContent.Name
 
-	if err != nil {
-		// Save failed. Now we have a snapshot asset outside of Kubernetes,
-		// but we don't have appropriate volumesnapshot content object for it.
-		// Emit some event here and controller should try to create the content in next sync period.
-		strerr := fmt.Sprintf("Error creating volume snapshot content object for snapshot %s: %v.", snapshotKey(snapshot), err)
-		klog.Error(strerr)
-		ctrl.eventRecorder.Event(newSnapshot, v1.EventTypeWarning, "CreateSnapshotContentFailed", strerr)
-		return nil, newControllerUpdateError(snapshotKey(snapshot), err.Error())
+		updateSnapshot, err := ctrl.clientset.SnapshotV1beta1().VolumeSnapshots(snapshot.Namespace).Update(snapshotCopy)
```

Comment:
this will not update the status of the snapshot? BoundVolumeSnapshotContentName needs to be updated by invoke UpdateStatus()
since Status will be updated later, this Update call mostly will not do anything. Is this still needed?